source/common/stats/stats_matcher_impl.cc

```diff
@@ -0,0 +1,49 @@
+#include "common/stats/stats_matcher_impl.h"
+
+#include <regex>
+#include <string>
+
+#include "common/common/utility.h"
+
+namespace Envoy {
+namespace Stats {
+
+StatsMatcherImpl::StatsMatcherImpl(const envoy::config::metrics::v2::StatsConfig& config) {
+  switch (config.stats_matcher().stats_matcher_case()) {
+  case envoy::config::metrics::v2::StatsMatcher::kInclusionList:
+    // If we have an inclusion list, we are being default-exclusive.
+    for (const auto& stats_matcher : config.stats_matcher().inclusion_list().patterns()) {
+      matchers_.push_back(Matchers::StringMatcher(stats_matcher));
+    }
+    is_inclusive_ = false;
+    break;
+  case envoy::config::metrics::v2::StatsMatcher::kExclusionList:
+    // If we have an exclusion list, we are being default-inclusive.
+    for (const auto& stats_matcher : config.stats_matcher().exclusion_list().patterns()) {
+      matchers_.push_back(Matchers::StringMatcher(stats_matcher));
+    }
+    FALLTHRU;
+  default:
+    // No matcher was supplied, so we default to inclusion.
+    is_inclusive_ = true;
+    break;
+  }
+}
+
+bool StatsMatcherImpl::rejects(const std::string& name) const {
+  //
+  //  is_inclusive_ | match | return
+  // ---------------+-------+--------
+  //        T       |   T   |   T     Default-inclusive and matching an (exclusion) matcher, deny.
+  //        T       |   F   |   F     Otherwise, allow.
+  //        F       |   T   |   F     Default-exclusive and matching an (inclusion) matcher, allow.
+  //        F       |   F   |   T     Otherwise, deny.
+  //
+  // This is an XNOR, which can be evaluated by checking for equality.
+
+  return (is_inclusive_ == std::any_of(matchers_.begin(), matchers_.end(),
```

Comment:
Do we want to short-circuit on `matchers_.empty()` to avoid creating the lambda when this feature isn't configured?

Reply:
I _think_ (please correct me if I'm wrong) that if this feature isn't configured, `matchers_` is empty, so `std::any_of()`'s iterator will just see that `matchers_.begin() == matchers_.end()` and avoid doing any work. Then, `std::any_of` defaults to false, which is the intended behavior here.

Reply:
I think that's correct. I was thinking of the instantiation of the lambda. But maybe that's optimized enough not to worry about.

Reply:
I'll come back to optimizing this if @jmarantz 's request for performance metrics ( https://github.com/envoyproxy/envoy/pull/4498#pullrequestreview-158763909 ) proves that this is inefficient.